Networking/WebSockets/WebSocketImpl.cc

```diff
@@ -369,17 +389,21 @@ namespace litecore { namespace websocket {
 
     // Initiates a request to close the connection cleanly.
     void WebSocketImpl::close(int status, fleece::slice message) {
-        if(!_didConnect && _framing) {
-            // The web socket is being requested to close before it's even connected, so just
-            // shortcut to the callback and make sure that onConnect does nothing now
-            closeSocket();
-            _closed = true;
-            
-            // CBL-1088: If this is not called here, it never will be since the above _closed = true
-            // prevents it from happening later.  This means that the Replicator using this connection
-            // will never be informed of the connection close and will never reach the stopped state
-            delegate().onWebSocketClose({kWebSocketClose, status, message});
-            return;
+        {
+            std::lock_guard<std::mutex> lock(_mutex);
+            if(!_didConnect && _framing) {
+
+                // The web socket is being requested to close before it's even connected, so just
+                // shortcut to the callback and make sure that onConnect does nothing now
+                closeSocket();
```

Comment:
IIRC it's troublesome to call `closeSocket()` -- an abstract virtual method -- while holding the lock, since the subclass implementation may do something that ends up calling back into this object. I don't remember specific cases, but none of the other calls to this method are within the scope of a `lock_guard`.